### PR TITLE
add: draft of md cell 12f1228a - ae9ee028

### PR DESCRIPTION
**Pull request by guixinliu**
@johnnychen94 前辈，我又翻译了些，请您 Review。

### CODE REVIEW

File: pluto.jl [hunk 1]
```diff
@@ -209,23 +202,30 @@ CPU 缓存旁边有一个叫做**预取器（prefetcher）**的小型电路。
 
 # ╔═╡ 12f1228a-8af0-11eb-0449-230ae20bfa7a
 md"""
-## Keep your data aligned to memory
-As just mentioned, your CPU will move entire cache lines of usually 512 consecutive bits (64 bytes) to and from main RAM to cache at a time. Your entire main memory is segmented into cache lines. For example, memory addresses 0 to 63 is one cache line, addresses 64 to 127 is the next, 128 to 191 the next, et cetera. Your CPU may only request one of these cache lines from memory, and not e.g. the 64 bytes from address 30 to 93.
+## 让数据在内存中保持对齐
+如上面所说，CPU 会一次性把 512 连续比特（即 64 字节）的“缓存线”数据从 RAM 移动到 CPU 缓存中。那么，主内存整体就会被相对分成一个又以一个缓存线。
```
Comment (johnnychen94) [inline pluto.jl:206]:
```suggestion
如上面所说，CPU 每次都是把整个缓存线上的 512 连续比特（即 64 字节）数据全部从主内存移动到 CPU 缓存中。那么，主内存整体就会被相对分成一个又以一个缓存线。
```

File: pluto.jl [hunk 2]
```diff
@@ -209,23 +202,30 @@ CPU 缓存旁边有一个叫做**预取器（prefetcher）**的小型电路。
...
+在接下来的例子中，我将使用指针在缓存线边界的给定偏移处重复访问数组。
+如果偏移位于范围 `0:56`，那么所有的整数都将位于一条缓存线上，函数运行起来也快。
+如果偏移位于范围 `57:63`，那么将会有整数穿过缓存线。
 """
 
 # ╔═╡ 3a1efd5a-8af0-11eb-21a2-d1011f16555c
-md"The consequences of unaligned memory access are very CPU-dependent. On my current CPU, I see a ~15% performance decrease. On my old computer where I originally wrote this notebook, the penalty was around 100%. Old processors can do [even worse things](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt) - incredibly, the CPU inside the Game Boy Advance from 2001 would _silently perform a different read!_ 😱
+md"未对齐内存访问的后果依 CPU 型号而定。在我现在的 CPU 上，会存在约 15% 的性能下降。
+ 在我最初写这本教程的旧电脑上，性能下降接近百分百。
+旧的处理器甚至会出现[更差的情况](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt) —— 难以置信的是，2001 年 Game Boy Advance 的 CPU 竟然会**静默执行不同的读取！** 😱
```
Comment (johnnychen94) [inline pluto.jl:221]:
```suggestion
旧的处理器甚至会出现[更差的情况](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt) —— 难以置信的是，2001 年 GBA 掌机上的 CPU 竟然会**悄无声息地读取两次！** 😱
```

@jakobnissen I'm unsure what "silently perform a diffrent read" means here. Does this mean the GBA CPU will execute two reads, or just execute one different read instruction?
It's also unclear to me what "read" means here -- does it mean"reading an unaligned 64bit integer", or "reading a cache line"?

File: pluto.jl [hunk 3]
```diff
@@ -250,22 +250,26 @@ end
...
+作为一名程序员，其实只有极少数情况会遇到内存对齐问题。这里提供两个例子：
 
-1. If you manually create an object with a strange size, e.g. by accessing a dense integer array with pointers. This can save memory, but will waste time. [My implementation of a Cuckoo filter](https://github.com/jakobnissen/Probably.jl) does this to save space.
-2. During matrix operations. If you have a matrix the columns are sometimes unaligned because it is stored densely in memory. E.g. in a 15x15 matrix of `Float32`s, only the first column is aligned, all the others are not. This can have serious effects when doing matrix operations: [I've seen benchmarks](https://juliasimd.github.io/LoopVectorization.jl/latest/examples/matrix_vector_ops/) where an 80x80 matrix/vector multiplication is 2x faster than a 79x79 one due to alignment issues.
+1. 若创建的对象具有奇怪的尺寸，例如使用指针访问稠密的整数数组。这种操作虽然可以节约内存，但是会浪费时间。我实现的[Cuckoo filter](https://github.com/jakobnissen/Probably.jl) 就使用了这种方式节约空间。 
+2. 矩阵操作的过程。因为数组元素在内存中紧靠着存储，所以有时数组的列会是未对齐的。例如，在一个 15x15 的 `Float32` 矩阵中，仅仅只有第一列是对齐的，其他列都存在未对齐问题。这会对矩阵操作造成严重的后果：[在链接里的基准测试中](https://juliasimd.github.io/LoopVectorization.jl/latest/examples/matrix_vector_ops/)，由于对齐问题，80x80 矩阵/向量的乘法会比 79x79 矩阵/向量的乘法快上两倍。
```
Comment (johnnychen94) [inline pluto.jl:259]: 单看中文好像不是很清楚 "紧靠着" 是什么意思

```suggestion
2. 矩阵操作的过程。因为数组元素在内存中是连续存储的，所以有时数组的列会是未对齐的。例如，在一个 15x15 的 `Float32` 矩阵中，仅仅只有第一列是对齐的，其他列都存在未对齐问题。这会对矩阵操作造成严重的后果：[在链接里的基准测试中](https://juliasimd.github.io/LoopVectorization.jl/latest/examples/matrix_vector_ops/)，由于对齐问题，80x80 矩阵/向量的乘法会比 79x79 矩阵/向量的乘法快上两倍。
```

File: pluto.jl [hunk 4]
```diff
@@ -283,13 +287,21 @@ end;
 
 # ╔═╡ ae9ee028-8af0-11eb-10c0-6f2db3ab8025
 md"""
-Let's break it down:
+让我们来分析一下：
```
Comment (johnnychen94) [inline pluto.jl:290]:
```suggestion
让我们来一点一点地分析它：
```

File: pluto.jl [hunk 5]
```diff
@@ -283,13 +287,21 @@ end;
...
 
-The lines beginning with `;` are comments, and explain which section of the code the following instructions come from. They show the nested series of function calls, and where in the source code they are. You can see that `eachindex`, calls `axes1`, which calls `axes`, which calls `size`. Under the comment line containing the `size` call, we see the first CPU instruction. The instruction name is on the far left, `movq`. The name is composed of two parts, `mov`, the kind of instruction (to move content to or from a register), and a suffix `q`, short for "quad", which means 64-bit integer. There are the following suffixes:  `b` (byte, 8 bit), `w` (word, 16 bit), `l`, (long, 32 bit) and `q` (quad, 64 bit).
+以 `;` 开头的行是注释，这些行会解释说明以下代码来自哪些部分。
```
Comment (johnnychen94) [inline pluto.jl:292]:
```suggestion
以 `;` 开头的行是注释，这些行会解释说明接下来的代码来自哪些部分。
```

File: pluto.jl [hunk 6]
```diff
@@ -283,13 +287,21 @@ end;
...
+它们会展示嵌套的函数调用，以及函数在源代码中的位置。
+你可以看到，`eachindex` 调用了 `axes1`，`axes1` 又调用了 `axes`，然后 `axes` 又调用了 `size`。
+在 `size` 那行注释下面，我们看到了第一条 CPU 指令。
+指令名位于最左侧，即 `movq`。
+名称由两部分组成，第一部分为 `mov`，它是指令的类型（即将数据移出或移入寄存器）；第二部分是后缀 `q`，这是 "quad" 的缩写，对应了 64-bit 整数。
+全部的后缀如下：`b` (byte, 8 位), `w` (word, 16 位), `l`, (long, 32 位) 和 `q` (quad, 64 位)。
 
-The next two columns in the instruction, `24(%rdi)` and `%rax` are the arguments to `movq`. These are the names of the registers (we will return to registers later) where the data to operate on are stored.
+指令中接着的两列，`24(%rdi)` 和 `%rax` 是 `movq` 的参数。
+它们都是存储待操作数据的寄存器的名字。后文将会详细讨论寄存器。
```
Comment (johnnychen94) [inline pluto.jl:301]:
```suggestion
它们都是存储待操作数据的寄存器的名字。不用担心，我们后面会再回过头来讲寄存器。
```

File: pluto.jl [hunk 7]
```diff
@@ -298,7 +310,9 @@ You can also see (in the larger display of assembly code) that the code is segme
      jne     L32
 ```
 
-The first instruction `cmpq` (compare quad) compares the data in registry `rdi`, which hold the data for the number of iterations left (plus one), with the number 1, and sets certain flags (wires) in the CPU based on the result. The next instruction `jne` (jump if not equal) makes a jump if the "equal" flag is not set in the CPU, which happens if there is one or more iterations left. You can see it jumps to `L32`, meaning this section repeat.
+第一条指令 `cmpq`（compare quad） 会将寄存器 `rdi` 中的数据与数字 1 进行比较，然后根据结果在 CPU 中设定一些 flag。
+其中，寄存器 `rdi` 中的数据对应了剩余的迭代次数（加 1）。
```
Comment (johnnychen94) [inline pluto.jl:314]: 👍

File: pluto.jl [hunk 8]
```diff
@@ -298,7 +310,9 @@ You can also see (in the larger display of assembly code) that the code is segme
...
+下一条指令 `jne`（jump if not equal， 如果不等则跳转）会在 CPU 未设置 "equal" flag 时发生跳转。即当还有一次或多次迭代未执行时，程序发生跳转。这条指令会跳转到 'L32' 节，这意味着会重复执行该部分。
```
Comment (johnnychen94) [inline pluto.jl:315]:
```suggestion
下一条指令 `jne`（jump if not equal， 如果不等则跳转）会在 CPU 未设置 "equal" flag 即迭代还没结束时发生跳转。这条指令会跳转到 'L32' 节，这意味着会重复执行该部分。
```

**Review by johnnychen94 — COMMENTED** [review]
There's one small part that I'm not sure what it means so maybe @jakobnissen can help explain it a bit?

File: pluto.jl (hunk 2 above)
Comment (jakobnissen) [reply pluto.jl:221]: I got it from here: https://fasterthanli.me/articles/peeking-inside-a-rust-enum
Apparently the hardware itself would not use the lowest 2 bits of the address when reading memory. Hence if you asked for some memory at `0x87654321` you would get the address at `0x87654320` instead.

**Pull request closed without merging**